File: challenges/triangles.rb
=begin

write a program to determine if a triangle is equilateral, isosceles or scalene.

equilateral has 3 sides, all the same langth
isosceles has two sides of same length
scalene has sides of different lentghts

all sides must be > 0 and the (sum of any two sides) > (third side)

based on the test cases, it looks like we create a triangle class, and pass in three sides
What are the methods?
=> kind - returns a string that tells us what kind of triangle we have
=> initialize, which will raise an ArgumentError for illegal triangles.
=end

class Triangle
  attr_reader :sides

  def initialize(s1, s2, s3)
    @sides = [s1, s2, s3]
    raise ArgumentError if invalid_sides?
  end

  def kind
    if number_of_equal_sides == 3
      "equilateral"
    elsif number_of_equal_sides == 2
      "isosceles"
    elsif number_of_equal_sides == 0
      "scalene"
    else
      "error"
    end
  end

  def number_of_equal_sides
    unique_value_array = sides.uniq
    unique_value_count = unique_value_array.size
    case unique_value_count
    when 1 then 3 # when we have 1 unique value, all 3 sides are equal
    when 2 then 2 # when we have 2 unique values, 2 sides are equal
    when 3 then 0 # when we have 3 unique values, then no sides are equal
    end
  end

  def invalid_sides?
    # make sure all sides are greater than 0
    return true if sides.include?(0)
    # make sure the sum rules apply
    return true if (sides[0] + sides[1]) <= sides[2]
    return true if (sides[0] + sides[2]) <= sides[1]
    return true if (sides[2] + sides[1]) <= sides[0]
    # return false if we've made it this far
    false
  end
end

=begin
(3, 4, 5)

3 + 4 > 5
3 + 5 > 4
4 + 5 > 3

=end

File: ls215/exercises/triangles.js
/*
write a function that takes three numbers as input. it tells us what kind of triangle the
three numbers make. either
  equilateral: all three sides are equal
  isosceles: two sides are equal
  scalene: all three sides are of a different length.
  invalid: see below.

an valid triangle has
  all sides greater than zero
  the sum of the two shortest sides must be greater than the sum of the longest side

this should be easy with an if else branch. have one function that returns boolena true
for each condition. start out with the invalid one first, then go to the other ones

Only tricky piece is how to find the longest side that's passed in.
could do this with a sort function: store the sides in an array, sort it and sum the 
first two elements and compare the sum to the last element.

What did LS do? 
  pretty much the same thing as me. Except they got the longest, middle and shortest sides
  using Math methods that I didn't even think about. 

  const perimeter = side1 + side2 + side3;
  const longest = Math.max(side1, side2, side3);
  const shortest = Math.min(side1, side2, side3);
  const middle = perimeter - longest - shortest;
*/

function isEquliateral(s1, s2, s3) {
  return (s1 === s2 && s2 === s3)
}

function isIsosceles(s1, s2, s3) {
  return (s1 === s2 || s2 === s3 || s1 === s3)
}

function isInvalid(s1,s2,s3) {
  //handle the 0 case
  if (s1 <= 0 || s2 <= 0 || s3 <= 0) {
    return true
  }
  let arr = [s1, s2, s3]
  arr.sort((ele1, ele2) => {
    if (ele1 < ele2) {
      return -1
    } else if (ele1 > ele2) {
      return 1
    } else if (ele1 === ele2) {
      return 0
    }
  })
  let largest = arr.pop()
  if (largest > arr[0] + arr[1]) {
    return true
  }
  return false
}

function triangle(s1, s2, s3) {
  if (isInvalid(s1, s2, s3)) {
    console.log("invalid");
    return "invalid"
  } else if(isEquliateral(s1, s2, s3)) {
    console.log("equilateral");
    return "equilateral"
  } else if (isIsosceles(s1,s2,s3)) {
    console.log("isosceles");
    return "isosceles"
  } else {
    console.log("Scalene");
    return "Scalene"
  }
}

//test cases
triangle(3, 3, 3);        // "equilateral"
triangle(3, 3, 1.5);      // "isosceles"
triangle(3, 4, 5);        // "scalene"
triangle(0, 3, 3);        // "invalid"
triangle(3, 1, 1);        // "invalid"
triangle(-1, 3, 3);        // "invalid"
triangle(1, 3, 1);        // "invalid"